src/compile/data/FilterInvalid.ts

```diff
@@ -0,0 +1,78 @@
+import {SCALE_CHANNELS} from '../../channel';
+import {isScaleChannel} from '../../channel';
+import {FieldDef} from '../../fielddef';
+import {hasContinuousDomain, scaleCompatible, ScaleType} from '../../scale';
+import {QUANTITATIVE, TEMPORAL} from '../../type';
+import {contains, Dict, differ, differArray, duplicate, extend, hash, keys, stringValue} from '../../util';
+import {VgFilterTransform, VgTransform} from '../../vega.schema';
+import {Model, ModelWithField} from '../model';
+import {UnitModel} from './../unit';
+import {DataFlowNode} from './dataflow';
+
+export class FilterInvalidNode extends DataFlowNode {
+  private filterInvalid: Dict<ScaleType>;
```

Comment:
As we can only have one dict, you can just reduce all of them together

model.reduceFieldDef((aggregator: Dict<FieldDef<string>>, fieldDef, channel)
and check if isScaleChannel(channel) should be pretty optimal.